text/separate-compilation.md

```diff
@@ -0,0 +1,229 @@
+- Title: Separate Compilation in Coq
+
+- Drivers: David Swasey, Paolo Giarrusso, Gregory Malecha
+
+# Summary
+
+Coq provides a module system that can be used explicitly through commands such
+as `Module` and `Module Type`. These can be quite heavyweight in many instances,
+and have some limitations when it comes to separately compiling files and
+building generic libraries.
+To address these problems, and support information hiding and Cardelli-style separate
+compilation, in this CEP we propose a notion of Coq interface files, which we
+will call a `.vi` file, and which are inspired by OCaml's `.mli` files.
+Intuitively, a Coq interface file called `module.vi` defines the public
+interface for `module.v`. The `module.vi` interface shall enable developing and
+typechecking clients, even before `module.v` has been implemented.
+
+If both `module.vi` and `module.v` are present, `module.vi` shall act as an
+opaque ascription for the top-level module defined by `module.v`; this opaque
+ascription ensures that clients that typecheck against `module.vi` shall still
+typecheck against the combination of `module.vi` and `module.v`, regardless of
+the implementation details of `module.v`, including any non-logical side effects
+such as hints (but excluding [universes](#universes)). Instead, without
+interfaces, adding implementations is sufficient to break clients, and changing
+implementations can break clients again, hindering modular development.
+
+This has a few advantages compared to Coq's state of the art:
+- It enables separate development: after agreeing on `module.vi`, `module.v` and
+  its clients can be developed independently. To ensure this, unlike today's
+  opaque ascription, `.vi` files can even hide side effects due to `Require`.
+- It reduces compile-time dependencies and improves compile times, even compared
+  to today's `vos` builds (initial builds can be more parallel, and incremental
+  builds need to recompile fewer files).
+
+NOTE: For concreteness, in this CEP we use the `.vi` extension for for interface
+source files, and the `.vio` extension for interface object files. However,
+`.vio` files should not be confused with `.vio` files produced by `-quick`
+builds: to avoid confusion, we could choose other file extensions or remove
+`-quick` builds entirely.
+
+# Proposed Semantics
+
+In this section, we describe the meaning of our Coq extension: instead of
+suggesting an implementation strategy, we sketch a semantics as a
+source-to-source transformation, to be taken as an informal and _approximate_
+specification. However, this is an approximate specification, because the exact
+semantics cannot be expressed via source-to-source transformation.
+
+As a key principle: `.vi` interfaces are meant to hide implementations and
```

Comment:
The use case for "implementations" is primarily proofs (plus necessary definitions)?

Reply:
`Qed` is typically good enough to hide proofs, but you often want to hide more: both hints needed by the implementation, but also the bodies of definitions, etc.

Reply:
i.e. same job as interfaces in java

Reply:
Not just, it can be useful to hide bodies of definitions that you do not want clients to think about though this does not work well if you want to compute on the definitions.